Plane: tailsitter add gain scaling options

kd0aij opened the pull request:
gain interpolation based on airspeed and no-airspeed gain based on attitude/throttle

new parameters: TAILSIT_SPDMIN,SPDMAX,GSCMSK,GSCMIN
SPDMIN/MAX control interpolation between FW and MC controllers
GSCMSK is a mask for enabling each of 3 gain scaling options: BOOST, ATT_THR, and INTERP
GSCMIN is the minimum scaling (range [0-1]) for the ATT_THR (attitude/throttle) based method
TAILSIT_THSCMX is the maximum gain boost applied at low throttle (default changed to 2)

@IamPete1 This is a refactored and rebased version of your gain interp method plus my att/throttle method and is close to what Pierre Losa has been flight testing. I was able to tune both new gain scaling methods to work well in SITL with the Stryker_quad.  I think this should be about ready to merge if you agree.

IamPete1 commented on ArduPlane/quadplane.cpp:
```diff
@@ -354,9 +354,9 @@ const AP_Param::GroupInfo QuadPlane::var_info2[] = {
+    // @Description: If > 1, maximum gain attenuation at high throttle/tilt angle, if > 1, max gain boost at low throttle
+    // @Range: 0.1 5
     // @User: Standard
     AP_GROUPINFO("TAILSIT_THSCMX", 3, QuadPlane, tailsitter.throttle_scale_max, 5),
```
might as well take the opportunity to change the default, I think 1.5 or 2 would be better.

kd0aij replied:
good point, 2 sounds good to me

IamPete1 commented on ArduPlane/tailsitter.cpp:
```diff
@@ -142,28 +189,73 @@ void QuadPlane::tailsitter_output(void)
+    if (aspeed > tailsitter.scaling_speed_min && have_airspeed && !is_zero(scaling_range)) {
+        // apply surface scaling to interpolate between fixed wing and VTOL outputs based on airspeed
+        // kd0aij note: that the airspeed estimate based only on GPS and (estimated) wind is
+        // not sufficiently accurate for tailsitters. (based on tests in RealFlight 8 with 10kph wind)
```
this is true, unless there is a airspeed sensor fitted, I guess we could check there is a actual airspeed sensor, but it is this airspeed that is used for control surface scaling in plane (i think).

kd0aij replied:
I figured the user would set spdmin=spdmax if there is no airspeed sensor

IamPete1 commented:
looks OK to me, i will do some testing in reaflight next week, I wonder if it would be worth having a scaling type pram rather than relying on 'magic' values might make it clearer whats going on. Maybe even a bitmask, i think my interpolation could work using your scailing,

kd0aij commented:
I forgot about the scaling type parameter... how about Q_TAILSIT_GSCL with values
'''
enum tailsitter_gain_scaling {
BOOST,
ATT_THROTTLE,
INTERP
}
It is tempting to go with a bitmask... I suppose all 3 methods could be active at once, and the code changes would be no more difficult... maybe?

IamPete1 commented:
looks good, i think it makes sense to have a bit mask, will simplify the code abit too, or at least make it more readable. All methods combined might work quite well.

kd0aij commented:
that was a fairly big change; I hope it's right
Tests OK in RF8 with the Stryker_quad, GSCMAK=ATT_THR+INTERP and GSCMIN=0.2
new logging record shows both the gain scaler and the VTOL_ratio:
![gain_scaling](https://user-images.githubusercontent.com/2300221/56756631-6a990000-674f-11e9-83e2-4625576cf152.png)

kd0aij commented on ArduPlane/tailsitter.cpp:
```diff
@@ -18,7 +18,58 @@
+// given b1 = -b0 = b: y_n = b (x_n - x_{n-1} - a y_{n-1}
+// x1 is the new input value, y is the filtered result
+// x0 is used to save the previous input
+inline void high_pass(float x1, float &x0, float &y) {
```
code style
default to no logging; build with logging for flight testing

kd0aij commented on ArduPlane/tailsitter.cpp:
```diff
@@ -142,28 +234,37 @@ void QuadPlane::tailsitter_output(void)
         }
     }
+
+    if (hal.util->get_soft_armed()) {
```
does this change behavior?

IamPete1 commented on ArduPlane/tailsitter.cpp:
```diff
@@ -56,79 +151,75 @@ void QuadPlane::tailsitter_output(void)
+        int32_t pitch_error_cd = (des_pitch_cd - ahrs_view->pitch_sensor) * 0.5;
         float extra_pitch = constrain_float(pitch_error_cd, -SERVO_MAX, SERVO_MAX) / SERVO_MAX;
         float extra_sign = extra_pitch > 0?1:-1;
         float extra_elevator = 0;
```
i would guess our Qasist bug is the extra elevator, it should be 
if (!is_zero(extra_pitch) && in_vtol_mode() )
this is because the ahrs_view-> pitch_sensor is only correct in vtol modes, i guess alternately you could still apply the extra elevator but use ahrs->pitch_sensor when not in a vtol mode

kd0aij replied:
fixed

kd0aij commented:
flight tested by @losawing: found a bug in roll limiting for bodyframe roll input modes
bug fixed in bc9dcd9

kd0aij commented:
@tridge I fixed the coding style and put the diagnostic logging behind a #define
Also fixed two bugs found during flight testing
Ready for a final review

kd0aij commented:
bugfixes flight tested by @losawing: https://discuss.ardupilot.org/t/dual-motor-tailsitters/15302/1799?u=kd0aij

> This morning I have loaded the last binaries in my batwing so non vectored dual motor and configured it as a copter tailsitter with q_tailsit_gscmsk=2 and max attenuation=0.2. I flown it with qacro mode and the result is absolutely amazing. The airframe is far better balanced than my test wing and the code run smoothly without oscillations at any speed and attitude, congratulation for that.

kd0aij commented:
rebased on master (after merge of #11673)

kd0aij commented:
@IamPete1 I've rebased and squashed this in preparation for trying to get it merged. Could you take a look to make sure your comments were all addressed? I've tried to keep this as close as possible to the branch that @losawing has flight tested on his quad+ copter tailsitters, and my SITL tests with the Stryker_quad model still look good.

kd0aij commented:
RealFlight8/SITL QACRO mode test of a 3D plane tailsitter looks pretty good IMHO even with a bad pilot: 
https://youtu.be/7UukEOiajIg

Gains set using QAUTOTUNE with
```
Q_A_ACCEL_P_MAX  100000.000000
Q_A_ACCEL_R_MAX  100000.000000
Q_A_ACCEL_Y_MAX  500000.000000
```

rmackay9 commented on libraries/AP_Motors/AP_MotorsMatrixTS.cpp:
```diff
@@ -140,6 +146,15 @@ void AP_MotorsMatrixTS::setup_motors(motor_frame_class frame_class, motor_frame_
+
+                    success = true;
+                    break;
+                case MOTOR_FRAME_TYPE_NYT_PLUS:
```
nitpick I'd put these cases in either numerical or alphabetical order

rmackay9 commented on libraries/AP_Motors/AP_Motors_Class.h:
```diff
@@ -59,6 +59,8 @@ class AP_Motors {
         MOTOR_FRAME_TYPE_DJI_X = 13, // X frame, DJI ordering
         MOTOR_FRAME_TYPE_CW_X = 14, // X frame, clockwise ordering
         MOTOR_FRAME_TYPE_I = 15, // (sideways H) octo only
+        MOTOR_FRAME_TYPE_NYT_PLUS = 16, // plus frame, no differential torque for yaw
```
Don't forget to add these to the FRAME_TYPE parameter description!  Perhaps with "(TailSitter only)" appended

IamPete1 commented on ArduPlane/tailsitter.cpp:
```diff
@@ -48,6 +103,51 @@ bool QuadPlane::tailsitter_active(void)
+                // due to reference frame change roll and yaw are swapped, use roll as rudder input and output direct as with plane
+                fw_aileron = plane.rollController.get_rate_out(-yaw_rate, speed_scaler);
+                fw_elevator = plane.pitchController.get_rate_out(pitch_rate, speed_scaler);
+                fw_rudder = plane.channel_roll->get_control_in();
```
I think this may not be correct with the new control options? Does the roll input always control tail-sitter roll?

kd0aij commented:
Which Realflight model are you using? I'm seeing good gain_interp results now with FT_Bronco_AV_Pete, except that rudder input is wrong in QACRO. 
Log here to show params and strange rudder behavior: https://drive.google.com/open?id=1IAnYPWhAZqeWcins-TYin-byGsuPT_YG
It responds to rudder with pitch changes. But the AddictionX is fine with similar parameters.
